### PR TITLE
[20.09] firefox: enable pipewire+webrtc for wayland users

### PR DESCRIPTION
**Pull request by worldofpeace**
(cherry picked from commit 31e54cdde34073b3b83e9fca4434cb0cedcb2829)

###### Motivation for this change
Backports https://github.com/NixOS/nixpkgs/pull/84233

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 1]
```diff
@@ -94,6 +94,11 @@ stdenv.mkDerivation ({
 
   patches = [
     ./env_var_for_system_dir.patch
+    (fetchpatch {
```
Comment (endgame) [inline pkgs/applications/networking/browsers/firefox/common.nix:97]: Should this only be in `patches` if `waylandSupport`?

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 2]
```diff
@@ -137,6 +142,11 @@ stdenv.mkDerivation ({
 
   postPatch = ''
     rm -rf obj-x86_64-pc-linux-gnu
+
+    # needed for enabling webrtc+pipewire
+    substituteInPlace \
```
Comment (endgame) [inline pkgs/applications/networking/browsers/firefox/common.nix:147]: Should this only run if `waylandSupport`?

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 3]
```diff
@@ -137,6 +142,11 @@ stdenv.mkDerivation ({
...
+      media/webrtc/trunk/webrtc/modules/desktop_capture/desktop_capture_generic_gn/moz.build \
+      --replace /usr/include ${pipewire.dev}/include
```
Comment (endgame) [inline pkgs/applications/networking/browsers/firefox/common.nix:149]: Do other libraries need their include dirs listed here?

**Review by endgame — COMMENTED** [review]
I don't know the firefox build system so some of these comments might be off-base.

**Comment by worldofpeace** [comment]
@endgame This is a backport. That means you really can't request changes to already committed code without encouraging diversion which has to be forward ported.

**Comment by endgame** [comment]
@worldofpeace Fair enough then.

**Comment by worldofpeace** [comment]
@endgame Not sure if you could submit a  review the original merged PR, though. Let's see what @andir thinks

**Comment by endgame** [comment]
I tried to build firefox from nixpkgs master without wayland and it broke because it couldn't find `-lpipewire` so  something needs to be reported: either on the original PR or as a new issue. No point pretending that wayland is an option if it breaks when turned off.

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 1 above)
Comment (andir) [reply pkgs/applications/networking/browsers/firefox/common.nix:97]: We should have less conditional patches as that makes it harder to have silent breakage. Usually those features should only be disabled via the build system and not via conditional patches.

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 3 above)
Comment (andir) [reply pkgs/applications/networking/browsers/firefox/common.nix:149]: Why would they?

Comment (andir) [reply pkgs/applications/networking/browsers/firefox/common.nix:149]: Looking at the patch we should probably also patch at least one other library:
```
+    CXXFLAGS += [ "-I/usr/include/pipewire-0.3" ]
+    CXXFLAGS += [ "-I/usr/include/spa-0.2" ]
```

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 1 above)
Comment (andir) [reply pkgs/applications/networking/browsers/firefox/common.nix:97]: After reading the patch, which doesn't support making pipewire optional, I am actually in favor of just dropping it or making it at least non-default. I am even less motivated to maintain this now then I was before looking at the patch. It seems like it needs to at least contain one more feature flag check in `media/webrtc/trunk/webrtc/modules/desktop_capture/desktop_capture_generic_gn/moz.build`. Probably the same that was applied in `media/webrtc/trunk/webrtc/modules/desktop_capture/BUILD.gn`. And `pipewire` and `spa` should be discovered via pkg_config (like in the other case) not just via hardcoded paths.

**Comment by andir** [comment]
Thank you for testing and reporting @endgame.

@worldofpeace due to the quality of the patch I didn't want it in the first place but was probably too inactive to make my opinion voiced enough. At this stage I'd at least refrain from adding it to the stable release. It is a nice feature and a pity but I do not see us having the manpower to do fix and support this patch.

I'll spend a bit of time trying to clean the patch up to fix the situation in both master and this backport. If that fails I'm still in favor of skipping it or making the entire code optional (which is a mess I really dislike).

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 3 above)
Comment (andir) [reply pkgs/applications/networking/browsers/firefox/common.nix:149]: Ok, these two both come from `pipewire` to this patching is actually correct.

**Comment by worldofpeace** [comment]
This patch is actually being used in more distros than I thought, see https://github.com/emersion/xdg-desktop-portal-wlr/wiki/Screencast-Compatibility#webrtc-aka-firefoxchromium

**Comment by andir** [comment]
> This patch is actually being used in more distros than I thought, see https://github.com/emersion/xdg-desktop-portal-wlr/wiki/Screencast-Compatibility#webrtc-aka-firefoxchromium

Yeah but most of them hide it behind a feature flag and/or only provide a single build for their entire distribution. Here we are trying to allow users to customize the package but didn't bake that into the derivation when this patch was added.

I just pushed c9c0b90b62bb863b56a7bdc356d38f7b143f2fa5 to this branch which does the required housekeeping for the added patch. I'll also add this to unstable once we agree on it.

**Comment by worldofpeace** [comment]
> > This patch is actually being used in more distros than I thought, see https://github.com/emersion/xdg-desktop-portal-wlr/wiki/Screencast-Compatibility#webrtc-aka-firefoxchromium
> 
> Yeah but most of them hide it behind a feature flag and/or only provide a single build for their entire distribution. Here we are trying to allow users to customize the package but didn't bake that into the expression when this patch was added.
> 
> I just pushed [c9c0b90](https://github.com/NixOS/nixpkgs/commit/c9c0b90b62bb863b56a7bdc356d38f7b143f2fa5) to this branch which does the required housekeeping for the added patch. I'll also add this to unstable once we agree on it.

Thanks, it looks good.

**Comment by worldofpeace** [comment]
@andir Could you also summarize your current stance on this? It seems you've taken a few and it's not clearly linear because some of it is in the review comments.

**Comment by andir** [comment]
This PR as is is probably fine. If you want to carry that into a release that sounds fine. I still worry about keeping the feature stable as we effectively now depend on Mozilla and a single fedora maintainer. Given the history with pipewire fedora will probably continue to provide patches. I just hope they'll do it on time as very often they were slower than us in updating the browser when security issues appeared.

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 2 above)
Comment (andir) [reply pkgs/applications/networking/browsers/firefox/common.nix:153]: That is what I added.

**Pull request merged**